ppafm/cpp/ProbeParticle.cpp

```diff
@@ -341,6 +345,118 @@ DLLEXPORT void getInPoints_LJ( int npoints, double * points_, double * FEs, int
+    const double k3  = -k[2];
+    const double k12 = -k1 * k2;
+
+    const double s6 = params[0];
+    const double s8 = params[1];
+    const double a1 = params[2];
+    const double a2 = params[3];
+
+    float L[MAX_REF_CN * MAX_REF_CN];
+
+    // Compute reference C6 weights for the probe particle
+    int pp_ind = elem_pp - 1;
+    double L_pp[MAX_REF_CN];
+    int i;
+    for (i = 0; i < MAX_REF_CN; i++) {
+        double pp_cn = ref_cn[pp_ind * MAX_REF_CN + i];
+        if (pp_cn >= 0.0f) { // Values less that 0 zero are invalid and should not be counted.
+            L_pp[i] = exp(k3 * pp_cn * pp_cn);
+        }
+    }
+    int max_ref_pp = i;
+
+
+    for (int ia = 0; ia < natoms; ia++) { // Loop over all atoms
+
+        const Vec3d pos = Ratoms[ia];
+        const int elem_ind = elems[ia] - 1;
+        const double r_cov_elem = r_cov[elem_ind];
+
+        // Compute the coordination number for this atom
+        double cn = 0;
+        for (int j = 0; j < natoms; j++) {
+            if (j == ia) continue; // No self-interaction for coordination number
+            double d = (Ratoms[j] - pos).norm();
+            double r = r_cov[elems[j] - 1] + r_cov_elem;
+            cn += 1.0f / (1.0f + exp(k12 * r / d + k1));
+        }
+
+        // Compute gaussian weights and normalization factor for all of the reference
+        // coordination numbers.
+        double norm = 0;
+        int a, b;
+        for (a = 0; a < MAX_REF_CN; a++) {
+            double ref_cn_a = ref_cn[elem_ind * MAX_REF_CN + a];
+            if (ref_cn_a < 0.0f) break; // Invalid values after this
+            double diff_cn_a = ref_cn_a - cn;
+            double L_a = exp(k3 * diff_cn_a * diff_cn_a);
+            for (b = 0; b < max_ref_pp; b++) {
+                double L_ab = L_a * L_pp[b];
+                norm += L_ab;
+                L[a * MAX_REF_CN + b] = L_ab;
+            }
+        }
+        int max_ref_a = a;
+
+        // If the coordination number is so high that the gaussian weights are all zero,
+        // then we put all of the weight on the highest reference coordination number.
+        if (norm == 0) {
+            int a_ind = (max_ref_a - 1) * MAX_REF_CN;
+            for (b = 0; b < max_ref_pp; b++) {
+                double L_ab = L_pp[b];
+                norm += L_ab;
+                L[a_ind + b] = L_ab;
+            }
+        }
+
+        // Compute C6 coefficient as a linear combination of reference C6 values
+        int n_zw = MAX_REF_CN * MAX_REF_CN;
+        int n_yzw = MAX_D3_ELEM * n_zw;
+        int pair_ind = elem_ind * n_yzw + pp_ind * n_zw;  // ref_c6 shape = (MAX_D3_ELEM, MAX_D3_ELEM, MAX_REF_CN, MAX_REF_CN)
+        double c6 = 0;
+        for (int a = 0; a < max_ref_a; a++) {
+            int a_ind = a * MAX_REF_CN;
+            for (int b = 0; b < max_ref_pp; b++) {
+                int L_ind = a_ind + b;
+                int c6_ind = pair_ind + L_ind;
+                c6 += L[L_ind] * ref_c6[c6_ind];
```

Comment:
Hmm, I'm perhaps not able to follow the algorithm, (I should read the article probably) ...

But the basic idea is that the more neighbors the atom have, the more polarizable it is, therefore the higher it the C6 coef? (Or am I completely off?) 

Does it consider shomehow that conjugated sp2 systems are much more delocalized than alifactic sp3 systems?

Reply:
There is a big table of reference values for the C6 coefficients for different coordination numbers. For every pair of elements there is a 5x5 matrix of values. For example, for H-C pair, the matrix is
```
[[ 7.2540984,  6.807746 ,  5.628885 ,  5.270787 ,  4.401504 ],
 [11.511539 , 10.789845 ,  8.820874 ,  8.245396 ,  6.769922 ],
 [ 0.       ,  0.       ,  0.       ,  0.       ,  0.       ],
 [ 0.       ,  0.       ,  0.       ,  0.       ,  0.       ],
 [ 0.       ,  0.       ,  0.       ,  0.       ,  0.       ]]
```
And there is a corresponding table of reference coordination number values for every element. For H and C:
```
H: [+0.9118, +0.0000, -1.0000, -1.0000, -1.0000]
C: [+0.0000, +0.9868, +1.9985, +2.9987, +3.9844]
```
Negative values are padding. You see that the non-negative values correspond to the non-zero rows and colums of the first matrix. The final C6 coefficient is a Gaussian interpolation of all of the valid reference C6 values.

This calculation corresponds to the set of equations (16) in the paper. I should probably write this into the wiki at some point.